Исправления по 15 спринту:

iliaalii opened the pull request:
- Подключение к БД и работа с ней
 - добавлены резервы предметов
 - добавлены комментарии к предметам, после выполненного резерва

iliaalii committed:
Исправления по 15 спринту:
 - Подключение к БД и работа с ней
 - добавлены резервы предметов
 - добавлены комментарии к предметам, после выполненного резерва

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/dao/BookingRepository.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.List;
+import java.util.Optional;
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от JpaRepository - мы уже объявляем бин.

iliaalii replied:
почистил

avfyodorov commented on src/main/java/ru/practicum/shareit/item/dao/CommentRepository.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Collection;
+import java.util.List;
+
+@Repository
```
Аналогично.
Здесь и в других интерфейсах. Посмотрите, пожалуйста.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,37 @@
+import ru.practicum.shareit.user.model.User;
+
+@UtilityClass
+public class BookingMapper {
```
Принято максимально избавляться от шаблонного кода, чтобы не перегружать и так большие и объемные проекты. Для маппинга сущностей часто используется библиотека Mapstruct, которая позволяет автоматически настраивать маппинг из одной сущности в другую с помощью описания интерфейсов и аннотаций.
Вот здесь есть отличный гайд для старта работы с библиотекой.
https://web.archive.org/web/20240619130052/https://rukovodstvo.net/posts/id_799/

//  уже нет. только из архива  https://rukovodstvo.net/posts/id_799/

Что такое Mapstruct и как правильно настроить его для модульного тестирования в SpringBoot приложениях
https://javarush.com/groups/posts/3698-chto-takoe-mapstruct-i-kak-praviljhno-nastroitjh-ego-dlja-moduljhnogo-testirovanija-v-springboo

При наличии свободного времени можно поизучать и попробовать прикрутить в проект
Необязательно, просто для информации .

iliaalii replied:
времени к сожалению особо нет -_-
но поизучал немного и решил добавить, в будующем точно упростит задачу по созданию
спасибо :)

avfyodorov commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -65,6 +65,18 @@ public ErrorResponse handleMissingHeader(MissingRequestHeaderException e) {
 
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleAvailabilityException(AvailabilityException e) {
```
И здесь тоже хорошо бы добавить логирование

avfyodorov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceDB.java:
```diff
@@ -0,0 +1,171 @@
+                    BookingDto nextBooking = null;
+
+                    if (item.getOwner().getId().equals(userId)) {
+                        lastBooking = bookingStorage
```
В этом месте, в цикле,  идут запросы к базе данных. Было бы хорошо избавиться от этих многочисленных запросов, которые сильно нагружают сервер. Хорошо бы переделать данный метод, чтобы уменьшить число обращений к БД .

avfyodorov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceDB.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public ItemDto find(Long userId, Long itemId) {
```
Transactional

для методов выборки лучше добавить параметр readonly=true

avfyodorov reviewed (commented):
Добрый день, Илья! 

В целом хорошо, правда, есть ряд уточнений.

iliaalii committed:
Исправления по 15 спринту:
 - Удалены лишние аннотации
 - Добавлена библиотека Mapstruct
 - Вынесены запросы к БД из циклов

iliaalii committed:
Исправления по 15 спринту:
 - Связал lombok и mapstruct

avfyodorov reviewed (commented):
Добрый день, Илья!

Замечаний нет.
Работа принята.
